core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -578,16 +579,23 @@ class LogManager(logDirs: Seq[File],
   }
 
   /**
-   * Make a checkpoint for all logs in provided directory.
-   */
-  private def checkpointLogRecoveryOffsetsInDir(dir: File): Unit = {
+    * Make a checkpoint for all logs in provided directory by default.
+    *
+    * @param dir the directory in which logs are checkpointed
+    * @param affectedLogs logs whose snapshots need to be cleaned. If it's None, the snapshot for all logs in the directory will be cleaned
+    */
+  // Only for testing
+  private[log] def checkpointLogRecoveryOffsetsInDir(dir: File, affectedLogs: Option[Seq[Log]] = None): Unit = {
```

Comment:
Thinking about this a bit more. I feel that Option[Seq[Log]] is a bit hard to understand. The issue is that we are trying to do 2 separate things--writing the recovery checkpoint file and deleting the snapshot---in a single method. Perhaps it's cleaner to split them into 2 separate methods. I am thinking of the following.

// write the recovery checkpoint in the provided directory
def checkpointLogRecoveryOffsetsInDir(dir: File)

// clean the producer snapshot files in the provided logs
def cleanSnapshot(logs: Seq[Log])

Would that be better?

Reply:
That's a good point Jun. My only concern is that we lose the fact that these two things should happen at the same time. Is that ever not true? Another option would be to replace `Option[Seq[Long]]` with `Seq[Log]` and then just have the caller always pass the sequence.

Reply:
It's mostly true that these two things should happen together. However, in the case when we want to delete a partition (asyncDelete), we just want to checkpoint the recovery offsets w/o the to be deleted partition. There is no need to delete the snapshot since the partition will be deleted.

But I agree, perhaps the better compromise is to have a single method like the following and force the caller to explicitly pass in the logs to clean the snapshots.

def checkpointRecoveryOffsetsAndCleanSnapshot(dir: File, logsToCleanSnapshot: Seq[Log])

Reply:
Sounds good.

Reply:
I guess it's a good idea to have this method split into two sub routines focusing on their own jobs.